src/coreclr/jit/assertionprop.cpp

```diff
@@ -5591,6 +5591,141 @@ void Compiler::optImpliedAssertions(AssertionIndex assertionIndex, ASSERT_TP& ac
+            !BasicBlock::sameEHRegion(target, switchBb))
+        {
+            // Same here, we still might be able to create some useful assertions, e.g. X != 0, etc.
+            continue;
+        }
+
+        AssertionInfo newAssertIdx = NO_ASSERTION_INDEX;
+
+        // Is this target a default case?
+        if (hasDefault && (jmpTargetIdx == jumpCount - 1))
+        {
+            // For default case we can create "X >= maxValue" assertion. Example:
+            //
+            //   void Test(ReadOnlySpan<byte> name)
+            //   {
+            //       switch (name.Length)
+            //       {
+            //           case 0: ...
+            //           case 1: ...
+            //           ...
+            //           case 7: ...
+            //           default: %name.Length is >= 8 here%
+            //       }
+            //
+            if (value > 0)
+            {
+                AssertionDsc dsc   = {};
+                dsc.assertionKind  = OAK_NOT_EQUAL;
+                dsc.op2.kind       = O2K_CONST_INT;
+                dsc.op2.vn         = vnStore->VNZeroForType(TYP_INT);
+                dsc.op2.u1.iconVal = 0;
+                dsc.op2.SetIconFlag(GTF_EMPTY);
+                if (vnStore->IsVNCheckedBound(opVN))
+                {
+                    // Create "arrBnd >= value" assertion
+                    dsc.op1.kind = O1K_CONSTANT_LOOP_BND;
+                    dsc.op1.vn   = vnStore->VNForFunc(TYP_INT, VNF_GE, opVN, vnStore->VNForIntCon(value));
+                }
+                else
+                {
+                    // Create "X u>= value" assertion
+                    dsc.op1.kind = O1K_CONSTANT_LOOP_BND_UN;
+                    dsc.op1.vn   = vnStore->VNForFunc(TYP_INT, VNF_GE_UN, opVN, vnStore->VNForIntCon(value));
+                }
+                newAssertIdx = optAddAssertion(&dsc);
+            }
+            else
+            {
+                continue;
+            }
+        }
+        else
+        {
+            // Create "VN == value" assertion.
+            AssertionDsc dsc   = {};
+            dsc.assertionKind  = OAK_EQUAL;
+            dsc.op1.lclNum     = BAD_VAR_NUM; // O1K_LCLVAR relies only on op1.vn in Global Assertion Prop
+            dsc.op1.vn         = opVN;
+            dsc.op1.kind       = O1K_LCLVAR;
+            dsc.op2.vn         = vnStore->VNForIntCon(value);
+            dsc.op2.u1.iconVal = value;
+            dsc.op2.kind       = O2K_CONST_INT;
+            dsc.op2.SetIconFlag(GTF_EMPTY);
+            newAssertIdx = optAddAssertion(&dsc);
+        }
+
+        if (newAssertIdx.HasAssertion())
+        {
+            // TODO-Cleanup: We shouldn't attach assertions to nodes in Global Assertion Prop.
+            // It limits the ability to create multiple assertions for the same node.
+            GenTree* tree = gtNewNothingNode();
+            fgInsertStmtAtBeg(target, fgNewStmtFromTree(tree));
```

Comment:
You should assert there are no PHI nodes in this block (there shouldn't be any since the block has a single pred edge).

Reply:
Added an assert